pyquil/paulis.py

```diff
@@ -677,6 +752,62 @@ def get_programs(self):
         coefficients = np.array([term.coefficient for term in self.terms])
         return programs, coefficients
 
+    def matrix(self, qubit_mapping={}, nqubits=None):
+        """Create the matrix representation of self.
+
+        :param nqubits:  Number of qubits to represent self on. If None, the maximum qubit from
+                         the PauliSum or qubit_mapping is used
+        :param qubit_mapping: A dictionary-like object that maps from :py:class`QubitPlaceholder`
+                              to :py:class:`int`
+        :return: np.matrix representing the PauliTerm
+        """
+
+        # check that qubit_map is valid
+        real_qubits = set()
+        placeholder_qubits = set()
+        qubits = self.get_qubits()
+        for q in qubits:
+            if isinstance(q, int):
+                real_qubits.add(q)
+            elif isinstance(q, QubitPlaceholder):
+                placeholder_qubits.add(q)
+            else:
+                raise ValueError(
+                    "the qubit {} is neither int nor QubitPlaceholder".format(q))
+
+        mapped_placeholders = set(qubit_mapping.keys())
+        mapped_qubits = set(qubit_mapping.values())
+
+        assert placeholder_qubits <= mapped_placeholders,\
+            "not all QubitPlaceholders are mapped to real qubits"
+        assert len(mapped_qubits) == len(mapped_placeholders),\
+            "qubit_mapping is not one to one"
+        assert len(mapped_qubits & real_qubits) == 0,\
+            "qubit_mapping maps to qubits already in use"
+
+        # check that nqubits makes sense
+        if nqubits is None:
+            nqubits = max(mapped_qubits | real_qubits) + 1
+        elif nqubits <= max(mapped_qubits | real_qubits):
+            raise ValueError(
+                "nqubits lower than the maximum qubit in the PauliTerm or qubit_map")
+
+        # create the matrix
+        matrix = np.zeros((2**nqubits, 2**nqubits), dtype=complex)
+        for term in self:
+            matrix += term.matrix(qubit_mapping=qubit_mapping,
+                                  nqubits=nqubits,
+                                  check_mapping=False)
+        return matrix
```

Comment:
Similar comment here as about the matrix rep for `PauliTerm` being unjustifiably large. So if my `PauliSum` were to only consist of qubits 1 and 3 for example, then I should get a 4 x 4 matrix rep, not a 16 x 16. On the other hand, even if my `PauliSum` consisted of at most 2-local terms (i.e. involving only 2 qubits each), but was on the whole operating on, say, 4 qubits, then I should have a 16 x 16 matrix rep. An example would be: `sX(0) * sY(2) + sZ(1) * sX(3)`

Reply:
Yup, I just realized that the solution as is isn't really satisfying. Unfortunately there is also the case, where the program acts on more qubits than the PauliSum. Think about having part of the system a heat bath and only wanting to measure a hamiltonian that acts on the other part of the system. I still need the full wavefunction and a matrix for the hamiltonian that has the same dimensions as the wavefunction.

So how about adding another optional argument, specifying which qubit indices to keep in the matrix (even though their indices might nowhere appear in the PauliSum)? This would kind of solve the problem, but at the cost cluttering the function call with too many arguments.

I mainly used this function to easily calculate expectation values from the wavefunctions I got from the `WavefunctionSimulator`. So I feel that the aim should be here that the matrix plays nicely along with the wavefunctions returned from the WavefunctionSimulator.

Reply:
I suppose part of the resolution here is identifying more precisely what the primary motivation for constructing the matrix representation is (for you, and future users). If we specify qubit indices, we run the risk of specifying qubits that might not actually be realized on a lattice, in which case we might also want to take in a `QuantumComputer` object as an additional input, at which point everything becomes even more cumbersome. 

I would again argue that such a function should instead simply focus on creating the max weight matrix representation of the `PauliSum`. If some `Program` were to actually operate in a larger Hilbert space, then perhaps the user could do something like `np.kron(heat_bath, other_system)` (to use your example).

As far as calculating expectation values using the `WavefunctionSimulator` goes, was it not sufficient to use `WavefunctionSimulator().expectation(prep_prog, pauli_terms)`?

Reply:
Here is, what I used the .matrix method for so far:
 - To calculate expectation values and(!) variances of these via `<wf|ham|wf>` and `<wf|ham**2> - <wf|ham|wf>**2`
 - To find the eigenvalues `E_i` and eigenvectors `psi_i` of my hamiltonians numerically via `numpy.linalg.eigh` and then calculate projections `<psi_i | wf>`
 - To visualize the energy spectrum of a hamiltonian and compare it to e.g. a density matrix by plotting both with `matplotlib.pyplot.matshow`

Now concerning the whole question of how to deal with qubits possibly not existing on the lattice, I will write something in a response to @karalekas first comment. But in general I guess we have to expect from the user to know, what lattice he is running on and keep the available qubits in mind when tensoring up his PauliSums.